### DISCUSSION TITLE
[14장] 불변 객체 성능 문제

### DISCUSSION THREAD

**Discussion by 0sunset0** [cowaine/is-my-code-that-strange#116] [Chapter 13-14 (모델링, 리팩터링)] [answered]
14장 리팩토링 코드 (+이전 장들 코드) 에는 클래스의 인스턴스 변수가 final로 선언되어 있습니다. 

```java
class Customer{
    final  PurchasePoint consumptionPoint;
}
```

이렇게 하면 변경하고 싶을 때 매번 새로운 객체를 만들어내야 하는데, 그럼 버려지는 객체가 많이 생길 것 같습니다.
불변 객체를 최적화하는 방법에는 무엇이 있을까요?

**Comment by Irisation23** [comment, marked as answer]
기본적으로 참조하지 않는 객체가 있다면 가비지 컬렉터가 일을 하지 않을까 하지만 
가비지 컬렉터가 일을 제대로 하지못할 정도의 어떤 이슈가 발생한다면 최적화 방법을 고려할 수 있다고 생각합니다.

많은 최적화 방법이 있지만 하나의 예로 들어본다면 `객체 풀링(Object Pooling)`  방법을 사용해 볼 수 있을 것 같습니다.

``` java
public final class PurchasePoint {
    private static final Map<Integer, PurchasePoint> CACHE = new HashMap<>();

    private final int points;

    private PurchasePoint(int points) {
        this.points = points;
    }

    public static PurchasePoint of(int points) { // 값으로 들어온 객체가 이미 존재한다면 CACHE에서 반환, 없다면 새로운 객체 반환
        return CACHE.computeIfAbsent(points, PurchasePoint::new);
    }

    public int getPoints() {
        return points;
    }
}
```

**Reply by corock** [reply]
객체 풀링이 뭔지 찾아보니 [플라이웨이트 패턴(_Flyweight Pattern_)](https://refactoring.guru/ko/design-patterns/flyweight) 으로 부르기도 하네요~

**Comment by 0sunset0** [comment]
https://velog.io/@jjinwo0/Study-%EC%9A%B0%EC%95%84%ED%95%9C%EC%8A%A4%ED%84%B0%EB%94%94-1%EC%A3%BC%EC%B0%A8-%ED%9A%8C%EA%B3%A0#%EC%B2%AB-%EC%8A%A4%ED%84%B0%EB%94%94-%ED%99%9C%EB%8F%99
